File: Leetcode/Graphs/733. Flood Fill.py
import queue

class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, newColor: int) -> List[List[int]]:
        def isValidPos(x, y, X, Y):
            return x >= 0 and x < X and y >= 0 and y < Y
        Queue = queue.LifoQueue()
        Queue.put((sr, sc))
        ds = [-1, 1, 0, 0]
        Visited = [[False for i in range(len(image[0]))] for j in range(len(image))]
        Visited[sr][sc] = True
        while Queue.qsize() > 0:
            (x, y) = Queue.get()
            for i in range(4):
                if isValidPos(x + ds[i], y + ds[3 - i], len(image), len(image[0])):
                    if not Visited[x + ds[i]][y + ds[3 - i]]:
                        if image[sr][sc] == image[x + ds[i]][y + ds[3 - i]]:
                            Visited[x + ds[i]][y + ds[3 - i]] = True
                            Queue.put((x + ds[i], y + ds[3 - i]))
                            image[x + ds[i]][y + ds[3 - i]] = newColor
        image[sr][sc] = newColor
        return image

File: Leetcode/Graphs/733. Flood Fill.cpp
class Solution {
private:
    bool isValidPos(int x, int y, int N, int M) {
        return x >= 0 && y >= 0 && x < N && y < M;
    }
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        queue <pair <int, int>> Queue;
        Queue.push({sr, sc});
        vector <int> ds = {1, -1, 0, 0};
        int N = image.size(), M = image[0].size();
        vector <vector <bool>> Visited(N, vector <bool> (M, false));
        Visited[sr][sc] = true;
        while (!Queue.empty()) {
            pair <int, int> Pos = Queue.front();
            Queue.pop();
            for (int i = 0; i < 4; i++) {
                int x = Pos.first + ds[i], y = Pos.second + ds[3 - i];
                if (isValidPos(x, y, N, M) && !Visited[x][y] && image[x][y] == image[sr][sc]) {
                    Queue.push({x, y});
                    image[x][y] = color;
                    Visited[x][y] = true;
                }
            }
        }
        image[sr][sc] = color;
        return image;
    }
};